web/pages/api/v0/bet/index.ts

```diff
@@ -1,24 +1,308 @@
-import { NextApiRequest, NextApiResponse } from 'next'
+import * as admin from 'firebase-admin'
+import { z } from 'zod'
 import {
-  CORS_ORIGIN_MANIFOLD,
-  CORS_ORIGIN_LOCALHOST,
-} from 'common/envs/constants'
-import { applyCorsHeaders } from 'web/lib/api/cors'
-import { fetchBackend, forwardResponse } from 'web/lib/api/proxy'
-
-export const config = { api: { bodyParser: false } }
-
-export default async function route(req: NextApiRequest, res: NextApiResponse) {
-  await applyCorsHeaders(req, res, {
-    origin: [CORS_ORIGIN_MANIFOLD, CORS_ORIGIN_LOCALHOST],
-    methods: 'POST',
-    allowedHeaders: 'Authorization',
+  DocumentReference,
+  FieldValue,
+  Query,
+  Transaction,
+} from 'firebase-admin/firestore'
+import { groupBy, mapValues, sumBy, uniq } from 'lodash'
+
+import { redeemShares } from 'functions/src/redeem-shares'
+import { Contract, CPMM_MIN_POOL_QTY } from 'common/contract'
+import { User } from 'common/user'
+import { FLAT_TRADE_FEE } from 'common/fees'
+import {
+  BetInfo,
+  getBinaryCpmmBetInfo,
+  getNewMultiBetInfo,
+  getNewMultiCpmmBetInfo,
+  getNumericBetsInfo,
+} from 'common/new-bet'
+import { addObjects, removeUndefinedProps } from 'common/util/object'
+import { LimitBet } from 'common/bet'
+import { floatingEqual } from 'common/util/math'
+import { filterDefined } from 'common/util/array'
+import { APIError } from 'common/api'
+import { log, validate } from 'web/lib/api/utils'
+import { newEndpoint } from 'web/lib/api/endpoint'
+
+export const config = { api: { bodyParser: true } }
+
+const placeBet = newEndpoint(async (req, userId: string) => {
+  log(`Inside endpoint handler for ${userId}.`)
+  const { amount, contractId } = validate(bodySchema, req.body)
+
+  const result = await firestore.runTransaction(async (trans) => {
```

Comment:
Do you happen to know what the typical latency between Vercel and Firestore is? That's another kind of potential pitfall with running this as a Vercel function, since we make many queries in serial. If latency between Vercel and Firestore were 50ms round trip, then that would be an increase of 30ms over the status quo. It seems that would make all binary contract bets take 100-200ms longer, which would be pretty sad. (Note that a round trip is necessary to start the transaction.)

If it's true that transaction retries are a big problem for us, increased latency inside the transaction will severely exacerbate that problem.